src/shmem_internal.h

```diff
@@ -246,4 +352,71 @@ static inline double shmem_internal_wtime(void) {
 long shmem_util_getenv_long(const char* name, int is_sized, long default_value);
 char *shmem_util_getenv_str(const char* name);
 
+#ifndef MAX
+#define MAX(A,B) (A) > (B) ? (A) : (B)
+#endif
+
+/* Language bindings helper macros */
+#define SHMEM_BIND_C_AMO(decl)                                  \
+  decl(int,      int,           SHM_INTERNAL_INT)               \
+  decl(long,     long,          SHM_INTERNAL_LONG)              \
+  decl(longlong, long long,     SHM_INTERNAL_LONG_LONG)
+
+#define SHMEM_BIND_C_EXTENDED_AMO(decl)                         \
+  SHMEM_BIND_C_AMO(decl)                                        \
+  decl(float,    float,         SHM_INTERNAL_FLOAT)             \
+  decl(double,   double,        SHM_INTERNAL_DOUBLE)
+
+#define SHMEM_BIND_C_INTS_OP(decl, s_op, op)                            \
+  decl(short,    short,         SHM_INTERNAL_SHORT,         s_op, op)   \
+  decl(int,      int,           SHM_INTERNAL_INT,           s_op, op)   \
+  decl(long,     long,          SHM_INTERNAL_LONG,          s_op, op)   \
+  decl(longlong, long long,     SHM_INTERNAL_LONG_LONG,     s_op, op)
+
+#define SHMEM_BIND_C_FLOATS_OP(decl, s_op, op)                          \
+  decl(float,    float,         SHM_INTERNAL_FLOAT,         s_op, op)   \
+  decl(double,   double,        SHM_INTERNAL_DOUBLE,        s_op, op)   \
+  decl(longdouble, long double, SHM_INTERNAL_LONG_DOUBLE,   s_op, op)
+
+#define SHMEM_BIND_C_CMPLX_OP(decl, s_op, op)                           \
+  decl(complexf, float complex, SHM_INTERNAL_FLOAT_COMPLEX, s_op, op)   \
+  decl(complexd, double complex,SHM_INTERNAL_DOUBLE_COMPLEX,s_op, op)
+
+#define SHMEM_BIND_F_RMA(decl)                                          \
+  decl(character, CHARACTER,    SIZEOF_FORTRAN_CHARACTER)               \
+  decl(complex,  COMPLEX,       SIZEOF_FORTRAN_COMPLEX)                 \
+  decl(double,   DOUBLE,        SIZEOF_FORTRAN_DOUBLE_PRECISION)        \
+  decl(integer,  INTEGER,       SIZEOF_FORTRAN_INTEGER)                 \
+  decl(logical,  LOGICAL,       SIZEOF_FORTRAN_LOGICAL)                 \
+  decl(real,     REAL,          SIZEOF_FORTRAN_REAL)
+
+#define SHMEM_BIND_F_SIZES(decl)        \
+  decl(mem,     MEM,    1)              \
+  decl(4,       4,      4)              \
+  decl(8,       8,      8)              \
+  decl(32,      32,     4)              \
+  decl(64,      64,     8)              \
+  decl(128,     128,    16)
```

Comment:
I guess with macros there is not much we can do with getting around these types of size defs...